In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import norm
import scipy.stats as st


## Case study 1

### About the Data set (Wine Quality.csv)
Attribute Information:

The dataset are related to red and white variants of the "Vinho Verde" wine. Vinho verde is a unique product from the Minho (northwest) region of Portugal. Medium in alcohol, is it particularly appreciated due to its freshness (specially in the summer). 

    1 - fixed acidity 
    2 - volatile acidity 
    3 - citric acid 
    4 - residual sugar 
    5 - chlorides 
    6 - free sulfur dioxide 
    7 - total sulfur dioxide 
    8 - density 
    9 - pH 
    10 - sulphates 
    11 - alcohol Output variable (based on sensory data): 
    12 - quality (score between 0 and 10)

## 1. Load the libraries:

In [5]:
data=pd.read_csv('wine_data.csv')

## 2. Importing the dataset:

In [6]:
data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[4898 rows x 12 columns]

## 3.EDA Descriptive Statistics:

### 3.1.  Measures of Central Tendency

Measures of Central Tendency define significant, representative and adequate values for a set of data, depending on what you want to analyze. They are the 

1. mean
2. median
3. quantiles and mode.

### 3.2 Check the  mean ,Mode, Median and Quantile for dataset?

#### Mean

In [7]:
data.mean()

fixed acidity             6.854788
volatile acidity          0.278241
citric acid               0.334192
residual sugar            6.391415
chlorides                 0.045772
free sulfur dioxide      35.308085
total sulfur dioxide    138.360657
density                   0.994027
pH                        3.188267
sulphates                 0.489847
alcohol                  10.514267
quality                   5.877909
dtype: float64

#### Mode

In [8]:
data.mode()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            6.8              0.28          0.3             1.2      0.044   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 29.0                 111.0    0.992  3.14        0.5   

   alcohol  quality  
0      9.4        6

#### Median and Quantile

### 3.4 check (Quantile 25%,50%,75%) for dataset

#### Quantile 25%

In [9]:
data.quantile(.25)

fixed acidity             6.300000
volatile acidity          0.210000
citric acid               0.270000
residual sugar            1.700000
chlorides                 0.036000
free sulfur dioxide      23.000000
total sulfur dioxide    108.000000
density                   0.991723
pH                        3.090000
sulphates                 0.410000
alcohol                   9.500000
quality                   5.000000
Name: 0.25, dtype: float64


#### Quantile 50%

In [10]:
data.quantile(.5)

fixed acidity             6.80000
volatile acidity          0.26000
citric acid               0.32000
residual sugar            5.20000
chlorides                 0.04300
free sulfur dioxide      34.00000
total sulfur dioxide    134.00000
density                   0.99374
pH                        3.18000
sulphates                 0.47000
alcohol                  10.40000
quality                   6.00000
Name: 0.5, dtype: float64

#### Quantile 75%

In [11]:
data.quantile(.75)

fixed acidity             7.3000
volatile acidity          0.3200
citric acid               0.3900
residual sugar            9.9000
chlorides                 0.0500
free sulfur dioxide      46.0000
total sulfur dioxide    167.0000
density                   0.9961
pH                        3.2800
sulphates                 0.5500
alcohol                  11.4000
quality                   6.0000
Name: 0.75, dtype: float64

## 4. Measures of Dispersion

Measures of Dispersion are measures that indicate how spread the data is, or how they vary. The measures of dispersion are range, variance, standard deviation and the absolute deviation, or mean absolute deviation.

### 4.1 check the Range,Variance,Standard Deviation,Absolute Deviation or Mean Absolute Deviation?

#### Range

In [12]:
data.max()-data.min()

fixed acidity            10.40000
volatile acidity          1.02000
citric acid               1.66000
residual sugar           65.20000
chlorides                 0.33700
free sulfur dioxide     287.00000
total sulfur dioxide    431.00000
density                   0.05187
pH                        1.10000
sulphates                 0.86000
alcohol                   6.20000
quality                   6.00000
dtype: float64

#### Variance

In [13]:
data.var()

fixed acidity              0.712114
volatile acidity           0.010160
citric acid                0.014646
residual sugar            25.725770
chlorides                  0.000477
free sulfur dioxide      289.242720
total sulfur dioxide    1806.085491
density                    0.000009
pH                         0.022801
sulphates                  0.013025
alcohol                    1.514427
quality                    0.784356
dtype: float64

#### Standard Deviation

In [14]:
data.std()

fixed acidity            0.843868
volatile acidity         0.100795
citric acid              0.121020
residual sugar           5.072058
chlorides                0.021848
free sulfur dioxide     17.007137
total sulfur dioxide    42.498065
density                  0.002991
pH                       0.151001
sulphates                0.114126
alcohol                  1.230621
quality                  0.885639
dtype: float64

#### Absolute Deviation or Mean Absolute Deviation

## 5. Covariance and Correlation

1. Covariance is a numerical measure that indicates the inter-dependency between two variables. 
2. Covariance indicates how two variables behave together in relation to their averages. 
3. A covariance of 0 indicates that the variables are totally independant, while a high and positive covariance value means that a variable is big when the other is big. Analogously, a negative covariance with a high absolute value means that one variable is big when the other is small.
4. Covariance can be calculated through the cov() function. 

### 5.1 Check the Covariance and Correlation scores . write your Observations?

#### Covariance

In [15]:
data.cov()

fixed acidity  volatile acidity  citric acid  \
fixed acidity              0.712114         -0.001931     0.029533   
volatile acidity          -0.001931          0.010160    -0.001823   
citric acid                0.029533         -0.001823     0.014646   
residual sugar             0.381022          0.032865     0.057829   
chlorides                  0.000426          0.000155     0.000302   
free sulfur dioxide       -0.708919         -0.166300     0.193630   
total sulfur dioxide       3.266013          0.382354     0.622989   
density                    0.000670          0.000008     0.000054   
pH                        -0.054265         -0.000486    -0.002992   
sulphates                 -0.001651         -0.000411     0.000861   
alcohol                   -0.125533          0.008400    -0.011278   
quality                   -0.084947         -0.017382    -0.000987   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity               0.381022   0.000426            -0.708919   
volatile acidity            0.032865   0.000155            -0.166300   
citric acid                 0.057829   0.000302             0.193630   
residual sugar             25.725770   0.009828            25.800578   
chlorides                   0.009828   0.000477             0.037674   
free sulfur dioxide        25.800578   0.037674           289.242720   
total sulfur dioxide       86.531303   0.184687           444.865891   
density                     0.012727   0.000017             0.014966   
pH                         -0.148684  -0.000298            -0.001587   
sulphates                  -0.015435   0.000042             0.114938   
alcohol                    -2.812740  -0.009684            -5.234509   
quality                    -0.438316  -0.004062             0.122878   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                     3.266013  0.000670 -0.054265  -0.001651   
volatile acidity                  0.382354  0.000008 -0.000486  -0.000411   
citric acid                       0.622989  0.000054 -0.002992   0.000861   
residual sugar                   86.531303  0.012727 -0.148684  -0.015435   
chlorides                         0.184687  0.000017 -0.000298   0.000042   
free sulfur dioxide             444.865891  0.014966 -0.001587   0.114938   
total sulfur dioxide           1806.085491  0.067352  0.014894   0.652645   
density                           0.067352  0.000009 -0.000042   0.000025   
pH                                0.014894 -0.000042  0.022801   0.002688   
sulphates                         0.652645  0.000025  0.002688   0.013025   
alcohol                         -23.476605 -0.002871  0.022565  -0.002448   
quality                          -6.576746 -0.000814  0.013297   0.005425   

                        alcohol   quality  
fixed acidity         -0.125533 -0.084947  
volatile acidity       0.008400 -0.017382  
citric acid           -0.011278 -0.000987  
residual sugar        -2.812740 -0.438316  
chlorides             -0.009684 -0.004062  
free sulfur dioxide   -5.234509  0.122878  
total sulfur dioxide -23.476605 -6.576746  
density               -0.002871 -0.000814  
pH                     0.022565  0.013297  
sulphates             -0.002448  0.005425  
alcohol                1.514427  0.474726  
quality                0.474726  0.784356

#### Correlation

In [16]:
data.corr()

fixed acidity  volatile acidity  citric acid  \
fixed acidity              1.000000         -0.022697     0.289181   
volatile acidity          -0.022697          1.000000    -0.149472   
citric acid                0.289181         -0.149472     1.000000   
residual sugar             0.089021          0.064286     0.094212   
chlorides                  0.023086          0.070512     0.114364   
free sulfur dioxide       -0.049396         -0.097012     0.094077   
total sulfur dioxide       0.091070          0.089261     0.121131   
density                    0.265331          0.027114     0.149503   
pH                        -0.425858         -0.031915    -0.163748   
sulphates                 -0.017143         -0.035728     0.062331   
alcohol                   -0.120881          0.067718    -0.075729   
quality                   -0.113663         -0.194723    -0.009209   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity               0.089021   0.023086            -0.049396   
volatile acidity            0.064286   0.070512            -0.097012   
citric acid                 0.094212   0.114364             0.094077   
residual sugar              1.000000   0.088685             0.299098   
chlorides                   0.088685   1.000000             0.101392   
free sulfur dioxide         0.299098   0.101392             1.000000   
total sulfur dioxide        0.401439   0.198910             0.615501   
density                     0.838966   0.257211             0.294210   
pH                         -0.194133  -0.090439            -0.000618   
sulphates                  -0.026664   0.016763             0.059217   
alcohol                    -0.450631  -0.360189            -0.250104   
quality                    -0.097577  -0.209934             0.008158   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                     0.091070  0.265331 -0.425858  -0.017143   
volatile acidity                  0.089261  0.027114 -0.031915  -0.035728   
citric acid                       0.121131  0.149503 -0.163748   0.062331   
residual sugar                    0.401439  0.838966 -0.194133  -0.026664   
chlorides                         0.198910  0.257211 -0.090439   0.016763   
free sulfur dioxide               0.615501  0.294210 -0.000618   0.059217   
total sulfur dioxide              1.000000  0.529881  0.002321   0.134562   
density                           0.529881  1.000000 -0.093591   0.074493   
pH                                0.002321 -0.093591  1.000000   0.155951   
sulphates                         0.134562  0.074493  0.155951   1.000000   
alcohol                          -0.448892 -0.780138  0.121432  -0.017433   
quality                          -0.174737 -0.307123  0.099427   0.053678   

                       alcohol   quality  
fixed acidity        -0.120881 -0.113663  
volatile acidity      0.067718 -0.194723  
citric acid          -0.075729 -0.009209  
residual sugar       -0.450631 -0.097577  
chlorides            -0.360189 -0.209934  
free sulfur dioxide  -0.250104  0.008158  
total sulfur dioxide -0.448892 -0.174737  
density              -0.780138 -0.307123  
pH                    0.121432  0.099427  
sulphates            -0.017433  0.053678  
alcohol               1.000000  0.435575  
quality               0.435575  1.000000

#### Observation:
    


If we consider quality as parameter We can't observe any corelation between other feature and Quality where as alcohol shows a fare bit good correlation .

### Example (Confidence Interval)

Hindustan Pencils Pvt. Ltd. is an Indian manufacturer of pencils, writing materials and other stationery items, established in 1958 in Mumbai. Nataraj brand of pencils manufactured by the company is expected to have a mean length of 172 mm and the standard deviation of the length is 0.02 mm.

To ensure quality, a sample is selected at periodic intervals to determine whether the length is still 172 mm and other dimensions of the pencil meet the quality standards set by the company.

You select a random sample of 100 pencils and the mean is 170 mm. 
Construct a 95% confidenct interval for the pencil length.

### Solution:

It is known that 
* $\overline{X}$ = 172 mm
* $\sigma$       = 0.02 mm
* n              = 100
We need to compute $\overline{X} \quad \pm 1.96 \frac {\sigma} {\sqrt{n}}$ 

In [2]:
ci     = 0.95
Xavg   = 172
sigma  = 0.02
n=100
s      = sigma / np.sqrt(n)
LCI, UCI = st.norm.interval(ci, loc = Xavg, scale = s) # Give confidence interval 95%, mean and std as arguments to get CI
print('95 %s confidence interval for population mean is %1.4f  to %1.4f' % ('%', LCI , UCI))


95 % confidence interval for population mean is 171.9961  to 172.0039


# The End